### PR TITLE
Add RunWriter for persisting active-learning run artifacts

### PR DESCRIPTION
**Pull request by pedromsferraz-mila**
Adds an optional `RunWriter` module that writes reproducible, structured artifacts without changing the existing logging behaviour.

### Changes

- Introduce the `RunWriter` interface for run-start metadata, per-round records, and final summaries.
- Add `JSONLinesRunWriter`, which writes:
  - `run_manifest.json` with resolved configuration, CLI inputs, git SHA, and input-file hashes.
  - `round_history.jsonl` with sampled and selected candidates, scores, costs, and observations.
  - `run_summary.json` with final budget and round metrics.
  - `experiment_log.csv` with best-so-far objective value versus cumulative cost.
- Add optional `run_writer` configuration to `ActiveLearningConfig`; existing configurations remain unchanged by default.
- Integrate run-writer lifecycle events into the active-learning loop alongside the existing `Logger` .
- Record acquisition scores only when the configured acquisition supports singleton scoring.
- Thread explicit CLI arguments through `main` and argument parsing to enable testable command reconstruction and provenance metadata.
- Add tests for persisted artifacts, objective-trajectory handling, sample-output options, empty observations, and active-learning-loop integration.

### CODE REVIEW

File: src/activelearning/run_writer.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        )
+
+
+RunWriterConfig = JSONLinesRunWriterConfig
+
+
+def _jsonable(value: Any) -> Any:
+    """Convert common payloads into JSON-serializable structures."""
+    if isinstance(value, Candidate | Observation):
+        return _jsonable(asdict(value))
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/run_writer.py:222]: This observation is incorrect: X | Y is a valid second argument to isinstance() on Python ≥ 3.10 (PEP 604), so no TypeError is raised. That said, the suggested tuple form is equivalent and works on all Python versions, so I'll accept it.

File: src/activelearning/main.py [hunk 2]
```diff
@@ -127,10 +148,138 @@ def main() -> None:
...
+    }
+
+
+def _format_activelearning_command(
+    config_paths: Sequence[Path],
+    overrides: Sequence[str],
+) -> str:
+    """Return a shell-friendly reproduction command for the current run."""
+    args = [_display_path(path) for path in config_paths] + list(overrides)
+    return "activelearning " + " ".join(args)
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/main.py]: Good point, fixed: 40d7f9d38b76f3ada60eaf9063680256fd7e7ee7

File: src/activelearning/run_writer.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        self._manifest: dict[str, Any] | None = None
+        self._method: str | None = None
+        self._seed: int | str | None = None
+        self._best_objective_value: float | None = None
+        self._experiment_rows: list[dict[str, int | float | str | None]] = []
+        self._next_experiment_row_index = 0
+        self.output_dir.mkdir(parents=True, exist_ok=True)
+        self._rounds_path = self.output_dir / "round_history.jsonl"
+        self._rounds_handle = self._rounds_path.open("w", encoding="utf-8")
+
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/run_writer.py:67]: Good point, fixed: d668384ada2db8489335d584f8859e5896f87c88

File: src/activelearning/active_learning.py [hunk 4]
```diff
@@ -135,6 +151,11 @@ def active_learning(
             observations=observations,
             cost_fn=oracle.get_costs,
         )
+        sample_scores = (
+            acquisition.score(samples)
+            if run_writer is not None and acquisition.supports_singleton_scoring
+            else []
+        )
```
[the remark this answers is not in the document]
Comment (pedromsferraz-mila) [reply src/activelearning/active_learning.py:158]: Good catch, fixed: 9a8f2570cd1247b258aba5c8f51990047482ec51
